feat(allocator): connect `Vec2` module and make it compile

Dunqing opened the pull request:
This PR aims to connect the Vec2 with the whole program and make it compile and pass the CI. Since the diff is hard to understand, we may need to read the commit one by one to understand the changes.

Dunqing commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#9647** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9647?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9646** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9646?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9647).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_allocator/src/vec2/raw_vec.rs:
```diff
@@ -720,6 +719,90 @@ fn capacity_overflow() -> ! {
+/// rather than directly invoking `panic!` or similar.
+///
+/// [`handle_alloc_error`]: ../../alloc/alloc/fn.handle_alloc_error.html
+unsafe fn realloc(
```
One thing is confusing me - where is this function's code copied from?

The doc comment looks like it came from here: https://github.com/fitzgen/bumpalo/blob/1d2fbea9e3d0c2be56367b9ad5382ff33852a188/src/alloc.rs#L322-L402

But the body of the function looks more like (but not exactly the same as) this one: https://github.com/fitzgen/bumpalo/blob/1d2fbea9e3d0c2be56367b9ad5382ff33852a188/src/lib.rs#L1917-L1936

Dunqing replied:
I am sorry this confused you. Yes, the comment and the function implementation are copied from these two places.

overlookmotel replied:
Ah ha OK. Thanks for clarifying.

overlookmotel commented:
@Dunqing I hope you don't mind, but I've split this PR up into 4 parts (the preceding 3 PRs are pulled out of this one), just so the diffs are easier to read.

Annoyingly Graphite erases all record of multiple commits when it merges a PR, so if you want to keep a "paper trail" without the diffs of different commits combined, they have to be in separate PRs.

The only substantial change I made was to disable the clippy lints which `bumpalo`'s code triggers (#9730), rather than fixing them all, as you had. Some of the changes related to pointers and arithmetic overflow are a bit subtle, and I don't 100% trust clippy's suggestions, so I thought it'd be better to defer making those changes.

I suggest that we get `Vec2` connected up, solve the lifetime problems, and check it all works before making changes to `bumpalo`'s code. Once we've got that far, we can re-enable lint rules and fix the warnings in batches - do the simple ones first, and then tackle the pointer-related ones in a separate PR - smaller diffs will make it easier to see what we're doing and check our work.

---

I've been through this PR in detail and verified that the changes to `alloc`, `dealloc` and `realloc` calls are equivalent to the originals. It's kind of weird that we have to use the `allocator_api2::Allocator` methods, but Bumpalo doesn't expose its internal `Alloc` trait, so I think you're right that this is the only solution.

So... if you're happy with the changes I've made, I think we can merge the stack up to and including this PR.

I'd like to take a closer look at `retain_mut` before we merge that one. And I think *maybe* I have a solution to the lifetime problems in #9656. But it's a bit of a brain-bender!

---

In case I've screwed anything up, all the commits from your original stack before I made any changes are backed up on [overlookmotel/vec-rewrite-original](https://github.com/oxc-project/oxc/tree/overlookmotel/vec-rewrite-original) branch.

Dunqing commented:
> @Dunqing I hope you don't mind, but I've split this PR up into 4 parts (the preceding 3 PRs are pulled out of this one), just so the diffs are easier to read.
> 
> Annoyingly Graphite erases all record of multiple commits when it merges a PR, so if you want to keep a "paper trail" without the diffs of different commits combined, they have to be in separate PRs.
> 


Thank you for splitting this PR, yes, this is also what I am thinking if I should do these different things in separate PRs. Currently, the whole stack is more clear about what I am doing, so of course I very much like you to do this.

> The only substantial change I made was to disable the clippy lints which `bumpalo`'s code triggers (#9730), rather than fixing them all, as you had. Some of the changes related to pointers and arithmetic overflow are a bit subtle, and I don't 100% trust clippy's suggestions, so I thought it'd be better to defer making those changes.
> 
> I suggest that we get `Vec2` connected up, solve the lifetime problems, and check it all works before making changes to `bumpalo`'s code. Once we've got that far, we can re-enable lint rules and fix the warnings in batches - do the simple ones first, and then tackle the pointer-related ones in a separate PR - a smaller diff in that PR will make it easier to see what we're doing and check our work. The pointer stuff being correct is critical for safety, so I personally I think it's worthwhile proceeding cautiously with that.
> 

Yes, I agree!

> I've been through this PR in detail and verified that the changes to `alloc`, `dealloc` and `realloc` calls are equivalent to the originals. It's kind of weird that we have to use the `allocator_api2::Allocator` methods, but Bumpalo doesn't expose its internal `Alloc` trait, so I think you're right that this is the only solution.
> 

Yes, this is why I forked Bumpalo and exposed the `alloc` module in the last experiment.

> So... if you're happy with the changes I've made, I think we can merge the stack up to and including this PR.
> 
> I'd like to take a closer look at `retain_mut` before we merge that one. And I think _maybe_ I have a solution to the lifetime problems in #9656. But it's a bit of a brain-bender!
> 
> In case I've screwed anything up, all the commits from your original stack before I made any changes are backed up on [overlookmotel/vec-rewrite-original](https://github.com/oxc-project/oxc/tree/overlookmotel/vec-rewrite-original) branch.

I am okay with all changes to this stack. The decision to click the merge button is up to you, merge it when you think the timing is right now.

overlookmotel commented:
### Merge activity

* **Mar 12, 10:43 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
feat(allocator): connect `Vec2` module and make it compile (#9647)

This PR aims to connect the Vec2 with the whole program and make it compile and pass the CI. Since the diff is hard to understand, we may need to read the commit one by one to understand the changes.

overlookmotel commented:
OK great. Merging the bottom 4 PRs now.
